File: Task/Calkin-Wilf-sequence/Nim/calkin-wilf-sequence.nim
type Fraction = tuple[num, den: uint32]

iterator calkinWilf(): Fraction =
  ## Yield the successive values of the sequence.
  var n, d = 1u32
  yield (n, d)
  while true:
    n = 2 * (n div d) * d + d - n
    swap n, d
    yield (n, d)

proc `$`(fract: Fraction): string =
  ## Return the representation of a fraction.
  $fract.num & '/' & $fract.den

func `==`(a, b: Fraction): bool {.inline.} =
  ## Compare two fractions. Slightly faster than comparison of tuples.
  a.num == b.num and a.den == b.den

when isMainModule:

  echo "The first 20 terms of the Calkwin-Wilf sequence are:"
  var count = 0
  for an in calkinWilf():
    inc count
    stdout.write $an & ' '
    if count == 20: break
  stdout.write '\n'

  const Target: Fraction = (83116u32, 51639u32)
  var index = 0
  for an in calkinWilf():
    inc index
    if an == Target: break
  echo "\nThe element ", $Target, " is at position ", $index, " in the sequence."

File: Task/Calkin-Wilf-sequence/Lua/calkin-wilf-sequence.lua
do  -- Calculate members of the Calkin-Wilf sequence

    local n, d = 0, 0

    local function firstCW()
        n, d = 1, 1
    end

    local function nextCW()
        n, d = d, ( d + ( 2 * d * ( n // d ) ) ) - n
    end

    io.write( "The first 20 elements of the Calkin-Wilf sequence are:\n    " )
    firstCW()
    for i = 1, 20 do
       io.write( " " .. n .. "/" .. d )
       nextCW()
    end
    io.write( "\n" )

    firstCW()
    local pos = 1
    while n ~= 83116 or d ~= 51639 do
       nextCW()
       pos = pos + 1
    end
    io.write( "Position of 83116/51639 in the sequence: " .. pos .. "\n" )

end
